<a href="https://colab.research.google.com/github/BluessyJazz/Programacion-Para-Ingenieria_2024-1/blob/main/SI02%20Numpy/BEDOYAa_SI02_v20240417.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## «*Olvidaba que en todo combate entre el fanatismo y el sentido común, pocas veces logra este último imponerse*».
### [Marguerite Yourcenar](https://es.wikipedia.org/wiki/Marguerite_Yourcenar)

# SI 02 NumPy

Actividad individual.

Resuelva los siguientes ejercicios con códigos que se ajusten al PEP8. Recuerde incluir los comentarios en el código y documentar las funciones con docstrings al estilo de Google.

**Importante:**

**no se permite la utilización de ciclos (for or while) para resolver los ejercicios.**

**todos los ejercicios deben ser resueltos con NumPy.**

## Ejercicio 00

Cree el arreglo_00 de 15 **columnas** y 1000 **filas** de la siguiente manera:

* cinco columnas con números aleatorios provenientes de una distribución normal de media 0 y desviación típica 1. Utilice como semilla el número de su documento de identidad sumado con el número de la columna.

* cinco columnas con números aleatorios enteros en los siguientes intervalos: 200 a 300, 400 a 600, 0 a 360, 100 a 900, y 1000 a 2000. Utilice como semilla el número de su documento de identidad.

* cinco columnas con números aleatorios binarios (0,1)

Muestre el resultado y las características del arreglo.

In [1]:
# Importar librerías necesarias
import numpy as np

# Verificar si colorama está instalado y si no, instalarlo
try:
    from colorama import Fore, Style
except ImportError:
    %pip install colorama
    from colorama import Fore, Style

# Semilla
seed = 1001140290

# Crear arreglo de 15 columnas y 1000 filas
arreglo_00 = np.zeros((1000, 15))

# Cinco columnas con números aleatorios provenientes de una distribución normal
#(media 0 y desviación típica 1) con semilla seed + número de columna

# Columna 1
np.random.seed(seed + 1)
arreglo_00[:, 0] = np.random.normal(0, 1, size=1000)

# Columna 2
np.random.seed(seed + 2)
arreglo_00[:, 1] = np.random.normal(0, 1, size=1000)

# Columna 3
np.random.seed(seed + 3)
arreglo_00[:, 2] = np.random.normal(0, 1, size=1000)

# Columna 4
np.random.seed(seed + 4)
arreglo_00[:, 3] = np.random.normal(0, 1, size=1000)

# Columna 5
np.random.seed(seed + 5)
arreglo_00[:, 4] = np.random.normal(0, 1, size=1000)

# Cinco columnas con números aleatorios enteros en los intervalos dados

# Semilla 
np.random.seed(seed)

# Columna 6, 7, 8, 9 y 10
arreglo_00[:, 5] = np.random.randint(200, 300, size=1000)
arreglo_00[:, 6] = np.random.randint(400, 600, size=1000)
arreglo_00[:, 7] = np.random.randint(0, 360, size=1000)
arreglo_00[:, 8] = np.random.randint(100, 900, size=1000)
arreglo_00[:, 9] = np.random.randint(1000, 2000, size=1000)

# Cinco columnas con números aleatorios binarios (0,1)
arreglo_00[:, 10:] = np.random.randint(0, 2, size=(1000, 5))

# Ajustar la configuración de impresión de NumPy
#np.set_printoptions(precision=18, suppress=True)
#np.set_printoptions()

# Mostrar el resultado y las características del arreglo

print(Fore.CYAN + Style.BRIGHT + "Arreglo 00:" + Style.RESET_ALL)
print(arreglo_00.round(2), "\n")

# Semilla aleatoria
np.random.seed()

# Imprimir una fila aleatoria
print(Fore.CYAN + Style.BRIGHT + "Fila aleatoria:" + Style.RESET_ALL)
print(arreglo_00[np.random.randint(0, 1000)], "\n")

# Características del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Características del arreglo:" + Style.RESET_ALL)
print(Fore.CYAN + Style.BRIGHT + 
      "Dimensión:" + Style.RESET_ALL, arreglo_00.ndim)
print(Fore.CYAN + Style.BRIGHT + 
      "Forma:" + Style.RESET_ALL, arreglo_00.shape)
print(Fore.CYAN + Style.BRIGHT + 
      "Tamaño:" + Style.RESET_ALL, arreglo_00.size)
print(Fore.CYAN + Style.BRIGHT + 
      "Tipo de dato:" + Style.RESET_ALL, arreglo_00.dtype)
print(Fore.CYAN + Style.BRIGHT + 
      "Tamaño en bytes:" + Style.RESET_ALL, arreglo_00.nbytes)


Arreglo 00:
[[-1.18 -0.78 -0.26 ...  1.    0.    0.  ]
 [ 2.67  0.81 -1.56 ...  1.    0.    1.  ]
 [-1.19 -2.13  0.02 ...  0.    1.    0.  ]
 ...
 [-1.8  -0.17 -2.16 ...  1.    0.    0.  ]
 [-2.2  -0.73 -1.52 ...  0.    0.    0.  ]
 [ 0.07 -0.7  -0.41 ...  1.    0.    1.  ]] 

Fila aleatoria:
[ 1.55670800e-01 -1.28541288e+00 -9.30518641e-01 -8.99651208e-01
 -6.43770315e-01  2.89000000e+02  5.52000000e+02  2.03000000e+02
  3.37000000e+02  1.93800000e+03  1.00000000e+00  1.00000000e+00
  0.00000000e+00  0.00000000e+00  1.00000000e+00] 

Características del arreglo:
Dimensión: 2
Forma: (1000, 15)
Tamaño: 15000
Tipo de dato: float64
Tamaño en bytes: 120000


## Ejercicio 01

Cree dos arreglos, de 15 columnas y 20 filas cada uno, a partir del arreglo_00.

Calcule el producto elemento a elemento entre los dos arreglos creados

Muestre el resultado y las características del arreglo que contiene el producto elemento a elemento.

In [2]:
# Importar librerías necesarias
import numpy as np
from colorama import Fore, Style

# Crear dos arreglos de 15 columnas y 20 filas a partir del arreglo_00
arreglo_01 = arreglo_00[:20, :].copy()
arreglo_02 = arreglo_00[20:40, :].copy()

# Calcular el producto elemento a elemento
producto = arreglo_01 * arreglo_02

# Mostrar el resultado y las características del arreglo
print(Fore.CYAN + Style.BRIGHT + "Arreglo Producto:" + Style.RESET_ALL)
print(producto.round(2), "\n")

# Características del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Características del arreglo:" + Style.RESET_ALL)
print(Fore.CYAN + Style.BRIGHT +
        "Dimensión:" + Style.RESET_ALL, producto.ndim)
print(Fore.CYAN + Style.BRIGHT +
        "Forma:" + Style.RESET_ALL, producto.shape)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño:" + Style.RESET_ALL, producto.size)
print(Fore.CYAN + Style.BRIGHT +
        "Tipo de dato:" + Style.RESET_ALL, producto.dtype)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño en bytes:" + Style.RESET_ALL, producto.nbytes)


Arreglo Producto:
[[-1.590000e+00  7.400000e-01  2.400000e-01 -2.510000e+00 -4.700000e-01
   7.362300e+04  2.493050e+05  5.700000e+04  1.133650e+05  2.511696e+06
   0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00]
 [ 1.940000e+00  1.360000e+00  3.430000e+00 -4.300000e-01  1.000000e-02
   4.949400e+04  2.847960e+05  7.007000e+03  5.339700e+04  3.848819e+06
   0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00  1.000000e+00]
 [ 2.610000e+00 -7.300000e-01  1.000000e-02 -1.300000e-01  1.700000e-01
   5.948800e+04  1.870140e+05  2.017800e+04  3.072960e+05  1.634400e+06
   0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00]
 [ 2.200000e-01 -9.100000e-01  1.670000e+00  2.800000e-01  3.100000e-01
   5.460700e+04  2.657200e+05  9.570000e+03  6.956250e+05  3.093552e+06
   0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00]
 [-4.000000e-02 -1.110000e+00 -9.000000e-01 -0.000000e+00  5.000000e-02
   6.707700e+04  2.357440e+05  1.668500e+0

## Ejercicio 02

Cree dos arreglos, de 10 columnas y 10 filas cada uno, a partir del arreglo_00.

Calcule el producto matricial entre los dos arreglos creados

Muestre el resultado y las características del arreglo que contiene el producto matricial.

In [3]:
# Importar librerías necesarias
import numpy as np
from colorama import Fore, Style

# Crear dos arreglos de 10 columnas y 10 filas a partir del arreglo_00 
arreglo_03 = arreglo_00[40:50, :10].copy()
arreglo_04 = arreglo_00[50:60, :10].copy()

# Calcular el producto matricial
producto_matricial = np.dot(arreglo_03, arreglo_04)

# Mostrar el resultado y las características del arreglo
print(Fore.CYAN + Style.BRIGHT + "Producto Matricial:" + Style.RESET_ALL)
print(producto_matricial.round(2), "\n")

# Características del arreglo
print(Fore.CYAN + Style.BRIGHT +
        "Características del arreglo:" + Style.RESET_ALL)
print(Fore.CYAN + Style.BRIGHT +
        "Dimensión:" + Style.RESET_ALL, producto_matricial.ndim)
print(Fore.CYAN + Style.BRIGHT +
        "Forma:" + Style.RESET_ALL, producto_matricial.shape)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño:" + Style.RESET_ALL, producto_matricial.size)
print(Fore.CYAN + Style.BRIGHT +
        "Tipo de dato:" + Style.RESET_ALL, producto_matricial.dtype)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño en bytes:" + Style.RESET_ALL, producto_matricial.nbytes)


Producto Matricial:
[[-5.28570000e+02 -3.87090000e+02  1.64577000e+03  2.33655000e+03
  -1.53919000e+03  7.23287150e+05  1.26876357e+06  6.06648130e+05
   1.94724651e+06  4.71809757e+06]
 [-4.85780000e+02 -2.48390000e+02  1.46496000e+03  2.28507000e+03
  -8.26690000e+02  8.53249290e+05  1.49902383e+06  7.06146890e+05
   2.29425705e+06  5.55784740e+06]
 [-5.90940000e+02 -2.42960000e+02  1.56843000e+03  2.42337000e+03
  -7.35470000e+02  8.33768440e+05  1.47115910e+06  6.98030530e+05
   2.27129140e+06  5.37612508e+06]
 [-4.24870000e+02 -3.07340000e+02  1.33948000e+03  1.88694000e+03
  -1.41791000e+03  6.58956270e+05  1.14362738e+06  5.37450100e+05
   1.75171554e+06  4.31866130e+06]
 [-5.88240000e+02  2.48500000e+01  1.13081000e+03  2.15919000e+03
   3.72940000e+02  7.84653430e+05  1.43570799e+06  6.42188440e+05
   2.10320566e+06  5.05178342e+06]
 [-4.06970000e+02 -1.41700000e+02  1.10820000e+03  1.86420000e+03
  -4.66690000e+02  7.28988960e+05  1.29338286e+06  5.93518070e+05
   1.93796438

## Ejercicio 03

Cree un arreglo, de 15 columnas y 10 filas, a partir del arreglo_00.

Calcule y muestre la suma de cada una de las columnas y de cada una de las filas.

In [4]:
# Importar librerías necesarias
import numpy as np
from colorama import Fore, Style

# Crear un arreglo de 10 filas y 15 columnas a partir del arreglo_00
arreglo_05 = arreglo_00[60:70, :].copy()

# Calcular la suma de las columnas y de las filas
suma_columnas = arreglo_05.sum(axis=0)
suma_filas = arreglo_05.sum(axis=1)

# Mostrar el resultado
print(Fore.CYAN + Style.BRIGHT + "Suma de las columnas:" + Style.RESET_ALL)
print(suma_columnas.round(2), "\n")
print(Fore.CYAN + Style.BRIGHT + "Suma de las filas:" + Style.RESET_ALL)
print(suma_filas.round(2), "\n")


Suma de las columnas:
[-2.110e+00  8.400e-01  3.060e+00  2.210e+00  4.280e+00  2.496e+03
  5.064e+03  1.844e+03  4.302e+03  1.473e+04  4.000e+00  6.000e+00
  3.000e+00  5.000e+00  1.000e+01] 

Suma de las filas:
[2982.83 3177.63 2516.03 2686.66 2915.27 3021.09 2806.05 3023.67 2449.7
 2893.34] 



## Ejercicio 04

Utilice el arreglo creado en el ejercicio anterior. Reconvierta el arreglo a uno de dimensión 5 columnas y 30 filas.

A partir de ese arreglo cree otro nuevo en el cual cada uno de los elementos de las filas es mayor que el promedio de la respectiva columna.

Muestre el arreglo obtenido y sus características.

In [5]:
# Importar librerías necesarias
import numpy as np
from colorama import Fore, Style

# Reconvertir el arreglo_05 a uno de 30 filas y 5 columnas
arreglo_06 = arreglo_05.copy().reshape(30, 5)

# Crear un arreglo con los elementos mayores que el promedio de la columna
arreglo_06_bool = arreglo_06 > arreglo_06.mean(axis=0)

# Valores del arreglo
arreglo_06_float = arreglo_06_bool * arreglo_06

# Mostrar el resultado booleano y las características del arreglo
print(Fore.CYAN + Style.BRIGHT + "Arreglo 07 (booleano):" + Style.RESET_ALL)
print(arreglo_06_bool[:5, :], "\n", "...")

# Características del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Características del arreglo:" + Style.RESET_ALL)
print(Fore.CYAN + Style.BRIGHT +
        "Dimensión:" + Style.RESET_ALL, arreglo_06_bool.ndim)
print(Fore.CYAN + Style.BRIGHT +
        "Forma:" + Style.RESET_ALL, arreglo_06_bool.shape)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño:" + Style.RESET_ALL, arreglo_06_bool.size)
print(Fore.CYAN + Style.BRIGHT +
        "Tipo de dato:" + Style.RESET_ALL, arreglo_06_bool.dtype)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño en bytes:" + Style.RESET_ALL, arreglo_06_bool.nbytes)

# Mostrar el arreglo con sus valores
print("\n" + Fore.CYAN + Style.BRIGHT + 
      "Arreglo 07 (valores):" + Style.RESET_ALL)
print(arreglo_06_float.round(2), "\n")

# Características del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Características del arreglo:" + Style.RESET_ALL)
print(Fore.CYAN + Style.BRIGHT +
      "Dimensión:" + Style.RESET_ALL, arreglo_06_float.ndim)
print(Fore.CYAN + Style.BRIGHT +
      "Forma:" + Style.RESET_ALL, arreglo_06_float.shape)
print(Fore.CYAN + Style.BRIGHT +
      "Tamaño:" + Style.RESET_ALL, arreglo_06_float.size)
print(Fore.CYAN + Style.BRIGHT +
      "Tipo de dato:" + Style.RESET_ALL, arreglo_06_float.dtype)
print(Fore.CYAN + Style.BRIGHT +
      "Tamaño en bytes:" + Style.RESET_ALL, arreglo_06_float.nbytes)


Arreglo 07 (booleano):
[[False False False False False]
 [ True  True  True  True  True]
 [False False False False False]
 [False False False False False]
 [ True  True False  True  True]] 
 ...
Características del arreglo:
Dimensión: 2
Forma: (30, 5)
Tamaño: 150
Tipo de dato: bool
Tamaño en bytes: 150

Arreglo 07 (valores):
[[  -0.   -0.    0.    0.    0.]
 [ 262.  509.  334.  770. 1104.]
 [   0.    0.    0.    0.    0.]
 [   0.    0.    0.   -0.    0.]
 [ 217.  544.    0.  569. 1783.]
 [   0.    0.    0.    0.    0.]
 [  -0.    0.    0.    0.    0.]
 [ 222.  517.  223.  255. 1292.]
 [   0.    0.    0.    0.    0.]
 [  -0.   -0.    0.    0.   -0.]
 [ 256.  503.   95.  144. 1687.]
 [   0.    0.    0.    0.    0.]
 [   0.   -0.    0.   -0.    0.]
 [ 293.  470.  155.  377. 1617.]
 [   0.    0.    0.    0.    0.]
 [   0.    0.    0.    0.    0.]
 [ 261.  406.  188.  675. 1485.]
 [   0.    0.    0.    0.    0.]
 [  -0.   -0.   -0.    0.    0.]
 [ 203.  460.   81.  368. 1692.]
 [   0.    0.

## Ejercicio 05

Utilice el arreglo creado en el ejercicio 03.

Ordene cada columna de forma descendente, de mayor a menor, de forma independiente. Luego cree un nuevo arreglo con la concatenación de todas las columnas ordenadas.

Muestre el resultado y las características del arreglo.

In [6]:
# Importar librerías necesarias
import numpy as np
from colorama import Fore, Style

# Ordenar cada columna de forma descendente
arreglo_07 = np.sort(arreglo_05, axis=0)[::-1]

# Mostrar el resultado y las características del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Arreglo 08 (columnas ordenadas concatendas):" + Style.RESET_ALL)
print(arreglo_07.round(2), "\n")

# Características del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Características del arreglo:" + Style.RESET_ALL)
print(Fore.CYAN + Style.BRIGHT +
        "Dimensión:" + Style.RESET_ALL, arreglo_07.ndim)
print(Fore.CYAN + Style.BRIGHT +
        "Forma:" + Style.RESET_ALL, arreglo_07.shape)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño:" + Style.RESET_ALL, arreglo_07.size)
print(Fore.CYAN + Style.BRIGHT +
        "Tipo de dato:" + Style.RESET_ALL, arreglo_07.dtype)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño en bytes:" + Style.RESET_ALL, arreglo_07.nbytes)


Arreglo 08 (columnas ordenadas concatendas):
[[ 1.860e+00  1.490e+00  8.900e-01  1.020e+00  1.500e+00  2.930e+02
   5.620e+02  3.340e+02  7.700e+02  1.783e+03  1.000e+00  1.000e+00
   1.000e+00  1.000e+00  1.000e+00]
 [ 9.800e-01  1.480e+00  8.900e-01  9.500e-01  1.450e+00  2.800e+02
   5.480e+02  3.200e+02  6.750e+02  1.692e+03  1.000e+00  1.000e+00
   1.000e+00  1.000e+00  1.000e+00]
 [ 8.000e-01  8.100e-01  7.200e-01  5.800e-01  1.100e+00  2.630e+02
   5.450e+02  2.620e+02  5.690e+02  1.687e+03  1.000e+00  1.000e+00
   1.000e+00  1.000e+00  1.000e+00]
 [ 2.200e-01  4.900e-01  4.600e-01  5.500e-01  6.400e-01  2.620e+02
   5.440e+02  2.230e+02  5.210e+02  1.617e+03  1.000e+00  1.000e+00
   0.000e+00  1.000e+00  1.000e+00]
 [ 1.300e-01 -1.700e-01  3.100e-01  4.700e-01  5.000e-01  2.610e+02
   5.170e+02  1.880e+02  5.070e+02  1.485e+03  0.000e+00  1.000e+00
   0.000e+00  1.000e+00  1.000e+00]
 [ 1.000e-02 -3.500e-01  2.900e-01  2.300e-01  4.200e-01  2.560e+02
   5.090e+02  1.550e+02  3.

## Ejercicio 06

Cree un arreglo, de 9 columnas y 9 filas, a partir del arreglo_00.

Agregue una nueva columna al final del arreglo que contenga la suma de cada una de las filas y otra columna que contenga el promedio de las nueve columnas originales del arreglo.

Muestre el resultado y las características del arreglo.

In [7]:
# Importar librerías necesarias
import numpy as np
from colorama import Fore, Style

# Crear un arreglo de 9 columnas y 9 filas a partir del arreglo_00
arreglo_09 = arreglo_00[70:79, :9].copy()

# Agregar una nueva columna con la suma de cada fila
suma_filas = arreglo_09.sum(axis=1)
arreglo_09 = np.column_stack((arreglo_09, suma_filas))

# Agregar una nueva columna con el promedio de las nueve columnas originales
promedio_columnas = arreglo_09[:, :-1].mean(axis=1)
arreglo_09 = np.column_stack((arreglo_09, promedio_columnas))


# Mostrar el resultado y las características del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Arreglo 09:" + Style.RESET_ALL)
print(arreglo_09.round(2), "\n")

# Características del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Características del arreglo:" + Style.RESET_ALL)
print(Fore.CYAN + Style.BRIGHT +
        "Dimensión:" + Style.RESET_ALL, arreglo_09.ndim)
print(Fore.CYAN + Style.BRIGHT +
        "Forma:" + Style.RESET_ALL, arreglo_09.shape)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño:" + Style.RESET_ALL, arreglo_09.size)
print(Fore.CYAN + Style.BRIGHT +
        "Tipo de dato:" + Style.RESET_ALL, arreglo_09.dtype)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño en bytes:" + Style.RESET_ALL, arreglo_09.nbytes)


Arreglo 09:
[[-8.20000e-01  1.02000e+00 -8.80000e-01 -3.90000e-01  5.30000e-01
   2.95000e+02  4.75000e+02  1.76000e+02  3.17000e+02  1.26245e+03
   1.40270e+02]
 [ 3.00000e-01 -6.70000e-01 -6.00000e-02  3.50000e-01  3.90000e-01
   2.77000e+02  4.91000e+02  3.05000e+02  4.39000e+02  1.51230e+03
   1.68030e+02]
 [ 1.21000e+00  1.30000e+00  4.80000e-01  1.75000e+00  2.50000e-01
   2.88000e+02  5.17000e+02  2.30000e+01  5.35000e+02  1.36800e+03
   1.52000e+02]
 [-1.50000e+00  6.00000e-02  8.50000e-01  3.30000e-01 -2.60000e-01
   2.66000e+02  5.98000e+02  2.91000e+02  4.71000e+02  1.62549e+03
   1.80610e+02]
 [-1.10000e-01 -1.38000e+00 -8.20000e-01  1.37000e+00  2.10000e-01
   2.00000e+02  4.40000e+02  2.25000e+02  2.78000e+02  1.14227e+03
   1.26920e+02]
 [ 3.70000e-01  1.33000e+00  3.00000e-01 -2.40000e-01  2.20000e-01
   2.66000e+02  4.42000e+02  3.52000e+02  1.21000e+02  1.18298e+03
   1.31440e+02]
 [ 1.60000e+00  1.68000e+00  5.30000e-01  5.50000e-01 -1.08000e+00
   2.13000e+02  4.750

## Ejercicio 07

A partir de la columna del arreglo 00 que contiene los números enteros entre 0 y 360, genere un nuevo arreglo de 10 x 100.

Genere un nuevo arreglo en el que cada valor corresponda a la conversión de grados sexagesimales a radianes del arreglo que acaba de crear

Muestre el resultado y las características del arreglo.

In [8]:
# Importar librerías necesarias
import numpy as np
from colorama import Fore, Style

# Crear un arreglo de 10 filas y 100 columnas 
# a partir de la columna 8 del arreglo_00
arreglo_10x100 = arreglo_00[:, 7].copy().reshape(10, 100)

# Arreglo 100 x 10 (100 filas, 10 columnas)
arreglo_100x10 = arreglo_10x100.copy().reshape(100, 10)

# Convertir los grados sexagesimales a radianes
#arreglo_10x100 = np.deg2rad(arreglo_10)

# Factor de conversión pi / 180
factor = np.pi / 180

# Multiplicar el arreglo por el factor de conversión
arreglo_10x100 = arreglo_10x100 * factor

# Arreglo 100 x 10 (100 filas, 10 columnas)
arreglo_100x10 = arreglo_10x100.copy().reshape(100, 10)


# Mostrar el resultado y las características del arreglo 10x100
print(Fore.CYAN + Style.BRIGHT + 
      "Arreglo 10x100:" + Style.RESET_ALL)
print(arreglo_10x100.round(2), "\n")

# Características del arreglo 10 x 100
print(Fore.CYAN + Style.BRIGHT + 
      "Características del arreglo:" + Style.RESET_ALL)
print(Fore.CYAN + Style.BRIGHT +
        "Dimensión:" + Style.RESET_ALL, arreglo_10x100.ndim)
print(Fore.CYAN + Style.BRIGHT +
        "Forma:" + Style.RESET_ALL, arreglo_10x100.shape)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño:" + Style.RESET_ALL, arreglo_10x100.size)
print(Fore.CYAN + Style.BRIGHT +
        "Tipo de dato:" + Style.RESET_ALL, arreglo_10x100.dtype)
print(Fore.CYAN + Style.BRIGHT +
        "Tamaño en bytes:" + Style.RESET_ALL, arreglo_10x100.nbytes)

# Mostrar el arreglo 100 x 10
print(Fore.CYAN + Style.BRIGHT + 
      "\nArreglo 100x10:" + Style.RESET_ALL)
print(Fore.CYAN + Style.BRIGHT +
        "Forma:" + Style.RESET_ALL, arreglo_100x10.shape)
print(arreglo_100x10.round(2), "\n")


Arreglo 10x100:
[[4.97 2.5  3.09 0.96 0.82 4.36 2.74 5.25 2.91 3.02 2.04 0.65 0.87 5.85
  2.97 0.28 6.21 3.19 4.56 4.01 3.49 0.86 1.99 3.04 6.2  2.57 3.47 0.93
  2.76 0.77 0.73 3.79 1.52 4.28 1.34 5.78 3.12 4.96 5.36 3.11 1.97 4.92
  2.91 2.79 2.5  4.54 3.04 4.52 3.07 0.63 2.79 4.12 5.59 4.5  4.75 0.02
  2.93 1.85 3.11 5.04 5.83 1.06 3.89 1.66 2.71 3.28 1.41 5.59 4.57 2.18
  3.07 5.32 0.4  5.08 3.93 6.14 0.56 3.94 1.47 2.39 4.68 5.67 5.52 2.3
  3.49 4.8  3.86 1.87 0.1  1.62 2.55 2.32 1.36 5.43 2.27 4.89 1.94 6.14
  0.89 1.15]
 [3.58 0.44 3.82 2.3  5.25 3.49 3.49 2.18 0.54 1.22 2.44 3.   1.94 4.45
  2.36 3.67 1.1  2.69 2.32 0.51 2.46 4.64 4.52 4.36 1.66 0.09 5.62 0.19
  0.66 3.51 1.29 3.56 5.29 3.42 1.06 2.81 3.68 0.99 2.44 2.32 1.76 4.52
  5.06 1.92 0.54 1.9  5.92 3.3  2.16 1.64 0.87 2.9  2.48 4.9  5.95 3.84
  4.15 2.79 0.42 1.95 5.11 3.3  0.28 3.26 3.51 1.12 0.79 1.31 5.95 3.82
  0.7  1.69 5.64 1.45 4.99 1.62 5.74 2.06 6.04 3.72 5.59 0.4  4.47 5.64
  4.64 4.   4.8  3.98 1.85 5.27 5.57

## Ejercicio 08

Cree un arreglo, de 15 columnas y 15 filas seleccionadas aleatoriamente (utilice el número de su documento de identidad como semilla), a partir del arreglo_00.

Calcule y muestre el determinante del arreglo, la transpuesta y la inversa.

In [9]:
# Importar librerías necesarias
import numpy as np
from colorama import Fore, Style

# Crear un arreglo de 15 columnas y 15 filas seleccionadas aleatoriamente
np.random.seed(1001140290)
arreglo_11 = arreglo_00[np.random.randint(0, 1000, 10), :10].copy()

# Calcular el determinante del arreglo
determinante = np.linalg.det(arreglo_11)

# Calcular la transpuesta del arreglo
transpuesta = arreglo_11.T

# Calcular la inversa del arreglo
inversa = np.linalg.inv(arreglo_11)

# Mostrar el resultado y las características del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Arreglo 11:" + Style.RESET_ALL)
print(arreglo_11.round(2), "\n")

# Determinante del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Determinante del arreglo:" + Style.RESET_ALL)
print(determinante, "\n")

# Transpuesta del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Transpuesta del arreglo:" + Style.RESET_ALL)
print(transpuesta.round(2), "\n")

# Inversa del arreglo
print(Fore.CYAN + Style.BRIGHT + 
      "Inversa del arreglo:" + Style.RESET_ALL)
print(inversa.round(2), "\n")


Arreglo 11:
[[-1.300e-01  1.500e-01 -1.080e+00 -9.400e-01  4.000e-01  2.530e+02
   4.200e+02  1.600e+01  1.240e+02  1.593e+03]
 [-4.300e-01  2.500e-01  1.000e-02 -1.620e+00  4.100e-01  2.500e+02
   5.260e+02  4.400e+01  5.930e+02  1.176e+03]
 [ 5.000e-02  1.100e-01  1.190e+00  8.900e-01  5.100e-01  2.790e+02
   4.630e+02  3.410e+02  2.680e+02  1.995e+03]
 [ 5.000e-02  4.700e-01  1.050e+00  8.600e-01  8.500e-01  2.110e+02
   4.040e+02  2.110e+02  5.050e+02  1.416e+03]
 [-7.000e-02  8.800e-01 -4.500e-01 -1.060e+00  1.700e-01  2.740e+02
   5.340e+02  1.000e+01  3.370e+02  1.086e+03]
 [ 2.070e+00 -8.300e-01  3.000e-01  1.030e+00  1.600e-01  2.390e+02
   4.210e+02  2.360e+02  8.800e+02  1.868e+03]
 [-3.100e-01  2.430e+00  9.700e-01  1.240e+00 -4.400e-01  2.940e+02
   4.030e+02  7.700e+01  3.580e+02  1.742e+03]
 [-1.950e+00 -2.500e+00  1.070e+00  8.700e-01 -4.600e-01  2.690e+02
   4.870e+02  1.620e+02  3.220e+02  1.112e+03]
 [-1.310e+00  1.530e+00 -4.700e-01  1.560e+00  2.600e-01  2.270e+02


## Ejercicio 09

Cree un arreglo, de 8 columnas y  8 filas, ambas seleccionadas aleatoriamente (utilice el número de su documento de identidad como semilla), a partir del arreglo_00.

Luego cree un arreglo de dimensión 1x8 con número reales aleatorios en el intervalo 0,10 (utilice el número de su documento de identidad como semilla).

Asuma que el primer arreglo representa los coeficientes de un sistema de ecuaciones lineales, y el segundo arreglo contiene las respectivas constantes. Encuentre la solución al sistema utilizando funciones de álgebra lineal de NumPy.

Muestre el valor de cada una de las variables que resuelve el sistema de ecuaciones.

In [10]:
# Importar librerías necesarias
# Librerías estándar de Python
from IPython.display import display, Latex

# Librerías de terceros
import numpy as np
from colorama import Fore, Style

# Crear un arreglo de 8 columnas y 8 filas seleccionadas aleatoriamente
np.random.seed(1001140290)
arreglo_12 = arreglo_00[np.random.randint(0, 1000, 8), :8].copy()

# Crear un arreglo de 1x8 con números reales aleatorios en el intervalo 0, 10
arreglo_13 = np.random.uniform(0, 10, size=8)

# Sistema de ecuaciones en LaTeX
print(Fore.CYAN + Style.BRIGHT + 
      "Sistema de ecuaciones:" + Style.RESET_ALL)
# Generar LaTeX para el sistema de ecuaciones
latex_str = r"\begin{align*}"

for i, row in enumerate(arreglo_12):
    equation = " + ".join([f"{coef:.2f}x_{{{j + 1}}}"
    for j, coef in enumerate(row) if coef != 0])
    equation += f" &= {arreglo_13[i]:.2f} \\\\"
    latex_str += equation
latex_str += r"\end{align*}"

display(Latex(latex_str))

# Resolver el sistema de ecuaciones lineales
solucion = np.linalg.solve(arreglo_12, arreglo_13)

# Mostrar el resultado
print(Fore.CYAN + Style.BRIGHT + 
      "Solución del sistema de ecuaciones:" + Style.RESET_ALL)
print(solucion.round(2), "\n")

# Generar LaTeX para mostrar la solución del sistema de ecuaciones
latex_solution = r"\begin{align*}"

for i, value in enumerate(solucion):
    latex_solution += f"x_{{{i+1}}} &= {value:.2f} \\\\"
latex_solution += r"\end{align*}"

display(Latex(latex_solution))


Sistema de ecuaciones:


<IPython.core.display.Latex object>

Solución del sistema de ecuaciones:
[ 2.92 -0.98  0.96  0.27  3.86  0.02  0.01 -0.04] 



<IPython.core.display.Latex object>

### Sistema de Ecuaciones
$$
\begin{align*}-0.13x_{1} + 0.15x_{2} + -1.08x_{3} + -0.94x_{4} + 0.40x_{5} + 253.00x_{6} + 420.00x_{7} + 16.00x_{8} &= 8.80 \\-0.43x_{1} + 0.25x_{2} + 0.01x_{3} + -1.62x_{4} + 0.41x_{5} + 250.00x_{6} + 526.00x_{7} + 44.00x_{8} &= 8.57 \\0.05x_{1} + 0.11x_{2} + 1.19x_{3} + 0.89x_{4} + 0.51x_{5} + 279.00x_{6} + 463.00x_{7} + 341.00x_{8} &= 0.05 \\0.05x_{1} + 0.47x_{2} + 1.05x_{3} + 0.86x_{4} + 0.85x_{5} + 211.00x_{6} + 404.00x_{7} + 211.00x_{8} &= 4.15 \\-0.07x_{1} + 0.88x_{2} + -0.45x_{3} + -1.06x_{4} + 0.17x_{5} + 274.00x_{6} + 534.00x_{7} + 10.00x_{8} &= 9.82 \\2.07x_{1} + -0.83x_{2} + 0.30x_{3} + 1.03x_{4} + 0.16x_{5} + 239.00x_{6} + 421.00x_{7} + 236.00x_{8} &= 7.73 \\-0.31x_{1} + 2.43x_{2} + 0.97x_{3} + 1.24x_{4} + -0.44x_{5} + 294.00x_{6} + 403.00x_{7} + 77.00x_{8} &= 3.54 \\-1.95x_{1} + -2.50x_{2} + 1.07x_{3} + 0.87x_{4} + -0.46x_{5} + 269.00x_{6} + 487.00x_{7} + 162.00x_{8} &= 0.30 \\\end{align*}
$$

### Solución del Sistema
$$
\begin{align*}x_{1} &= 2.92 \\x_{2} &= -0.98 \\x_{3} &= 0.96 \\x_{4} &= 0.27 \\x_{5} &= 3.86 \\x_{6} &= 0.02 \\x_{7} &= 0.01 \\x_{8} &= -0.04 \\\end{align*}
$$